integration/experiment/power_sweep/gen_policy_recommendation.py

```diff
@@ -214,7 +257,75 @@ def policy_min_energy(plrange, enmodel, rtmodel = None, pltdp = None,
                 'energy': best_energy}
 
 
-def main(full_df, region_filter, dump_prefix, min_pl, max_pl, tdp, max_degradation):
+def rootssq(ll):
+    "Return the root of the sum of the squares of the list of values."
+    try:
+        return sum([ss**2  for ss in ll])**0.5
+    except TypeError:
+        pass
+    return ll
+
+def normal_comparison(dist1, dist2):
+    """Returns the probability that a sample drawn from the normal distribution
+    described by dist1 (a tuple of mu and sigma) will be larger than a sample
+    drawn from a normal distribution drawn from the normal distribution
+    described by dist2."""
+    mu1, sigma1 = dist1
+    mu2, sigma2 = dist2
+    sigma1 = rootssq(sigma1)
+    sigma2 = rootssq(sigma2)
+    return 0.5 * (1 + math.erf((mu1 - mu2) / (2 * (sigma1 ** 2 + sigma2 ** 2)**0.5)))
+
+
+def policy_confident_energy(plrange, enmodel, rtmodel = None, pltdp = None,
+                            max_degradation = None, confidence = 0.9):
+    """
+    Find the power limit over the range plrange (list-like) that
+    has the minimum predicted energy usage (according to the energy model
+    enmodel, an instance of PowerLimitModel), subject to the constraint
+    that its runtime does not exceed the runtime at power limit pltdp
+    by more than a factor of (1 + max_degradation), if max_degradation is
+    specified, with at least a confidence (default 0.9) probability. Returns a
+    dictionary with keys power, runtime, and energy, and values the optimal
+    power limit, the predicted runtime and energy at that limit,
+    respectively, and the standard deviations of these values."""
+    if pltdp is None:
+        pltdp = max(plrange)
+
+    en_predictions = [enmodel.evaluate(pl, True) for pl in plrange]
+    if max_degradation is None:
+        # we don't need the runtime model in this case
+        best_energy, best_pl = min(zip(en_predictions, plrange))
+        if rtmodel:
+            best_runtime = rtmodel.evaluate(best_pl, True)
+        else:
+            best_runtime = None
+        return {'power': best_pl,
+                'runtime': best_runtime[0],
+                'runtimedev': rootssq(best_runtime[1]),
```

Comment:
rootssq != standard deviation, you need to remove the mean.  Has the mean been removed from the best_runtime?

Reply:
To make this clearer, when the evaluate() method is called, each of the three returned values should get a different variable name.  This will avoid the confusion I had when thinking that "best_runtime[1]" is a container with runtime values in it.